syntax="proto3";//protobuf的版本

package fixbug;//声明代码所在的包 相当于namespace  后面通过protoc命令生成c++源码 fixbug就是他的namespace

//定义下面的选项，表示生成service服务类和rpc方法描述，默认不生成
option cc_generic_services=true;

message ResultCode
{
    int32 errcode=1;
    bytes errmsg=2;
}

//数据 列表 映射表
//定义登陆消息类型 name pwd
message  LoginReques
{
    // string name=1;//1表示第一个字段
    // string pwd=2;//第2个字段
    bytes name=1;//1表示第一个字段 一般把string->定义成bytes
    bytes pwd=2;//第2个字段
}

//定义登陆响应消息类型
message LoginResponse
{
    // int32 errcode=1;//消息响应类型 第1个字段
    // bytes errmsg=2;//一般把string->定义成bytes
    ResultCode result=1;//result 封装了上面的errcode和errmsg
    bool success=2;//登陆是否成功 第3个字段
}


message GetFriendListsRequest
{
    uint32 userid=1;

}

//返回的用户的信息
message User
{
    bytes name=1;
    uint32 age=2;
    enum Sex
    {
        MAN=0;
        WOMAN=1;
    }
    Sex sex=3;
}

message GetFriendListsResponse
{
    // int32 errcode=1;
    // bytes errmsg=2;
    ResultCode result=1;
    repeated User friend_list=2;//repeated表示是个好友列表类型
}

//在protobuf里面定义描述rpc方法的类型 -service
service UserServiceRpc
{
    rpc Login (LoginReques) returns (LoginResponse);
    rpc GetFriendLists (GetFriendListsRequest) returns (GetFriendListsResponse);

}//会生成一个服务提供者Callee ServieProvider（Class UserServiceRpc类）  
//及Caller ServiceConsumer rpc服务消费者(Class UserServiceRpc_Stub)
//UserServiceRpc_Stub(::PROTOBUF_NAMESPACE_ID::RpcChannel* channel); 没有默认构造函数
//UserServiceRpc_Stub:public UserServiceRpc